DashAI/back/plugins/utils.py

```diff
@@ -68,3 +76,79 @@ def get_plugins_from_pypi() -> List[dict]:
         if plugin_name.lower().startswith("dashai") and plugin_name.lower() != "dashai"
     ]
     return [_get_plugin_by_name_from_pypi(plugin_name) for plugin_name in plugins_names]
+
+
+def available_plugins() -> List[type]:
+    """
+    Get available DashAI plugins entrypoints
+
+    Returns
+    ----------
+    List[type]
+        A list of plugins' classes
+    """
+    # Retrieve plugins groups (DashAI components)
+    plugins = entry_points(group="dashai.plugins")
+
+    # Look for installed plugins
+    plugins_list = []
+    for plugin in plugins:
+        # Retrieve plugin class
+        plugin_class = plugin.load()
+        plugins_list.append(plugin_class)
+
+    return plugins_list
+
+
+def register_new_plugins(component_registry) -> List[type]:
+    """
+    Register only new plugins in component registry
+
+    Parameters
+    ----------
+    component_registry : ComponentRegistry
+        The current app component registry
+
+    Returns
+    ----------
+    List[type]
+        A list of plugins' classes that were registered in the component registry
+    """
+    installed_plugins = []
+    new_plugins = []
+    for component in component_registry.get_components_by_types():
+        installed_plugins.append(component["class"])
+    for plugin in available_plugins():
+        if plugin not in installed_plugins:
+            new_plugins.append(plugin)
+            # The component shouldnt be registered if it does not inherit from
+            # any DashAI base class with a 'TYPE' class attribute.
+            with contextlib.suppress(TypeError):
+                component_registry.register_component(plugin)
+
+    return new_plugins
+
+
+def install_plugin_from_pypi(pypi_plugin_name: str):
+    """
+    Register only new plugins in component registry
+
+    Parameters
+    ----------
+    pypi_plugin_name : str
+        A string with the name of the plugin in pypi to install
+
+    Raises
+    ------
+    RuntimeError
+        If pip install command fails
+    """
+    res = subprocess.run(
+        ["pip", "install", pypi_plugin_name],
+        stderr=subprocess.PIPE,
+        text=True,
+    )
+    if res.returncode != 0:
+        errors = [line for line in res.stderr.split("\n") if "ERROR" in line]
+        error_string = "\n".join(errors)
+        raise RuntimeError(error_string)
```

Comment:
Just to comment: what happens if this exception occurs at runtime - does the server die?

Reply:
No, but im not sure where the Error is being catched. Here is the output:

INFO:     127.0.0.1:62601 - "POST /plugin/index HTTP/1.1" 201 Created
ERROR:DashAI.back.dependencies.database.sqlite_database:Session rollback: An exception was raised while trying to operate the database.
Traceback (most recent call last):
  File "C:\Users\lucas\OneDrive\Desktop\DashAI\DashAI\back\dependencies\database\sqlite_database.py", line 51, in session
    yield session
  File "C:\Users\lucas\OneDrive\Desktop\DashAI\DashAI\back\api\api_v1\endpoints\plugins.py", line 256, in update_plugin
    install_plugin_from_pypi(plugin_name)
  File "C:\Users\lucas\OneDrive\Desktop\DashAI\DashAI\back\plugins\utils.py", line 158, in install_plugin_from_pypi
    raise RuntimeError(error_string)
RuntimeError: ERROR: Could not find a version that satisfies the requirement paquete_no_existente (from versions: none)
ERROR: No matching distribution found for paquete_no_existente
INFO:     127.0.0.1:62600 - "PATCH /plugin/1 HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\uvicorn\protocols\http\httptools_impl.py", line 419, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\uvicorn\middleware\proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\fastapi\applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\middleware\errors.py", line 184, in __call__
    raise exc
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\middleware\errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\middleware\cors.py", line 91, in __call__
    await self.simple_response(scope, receive, send, request_headers=headers)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\middleware\cors.py", line 146, in simple_response
    await self.app(scope, receive, send)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\middleware\exceptions.py", line 79, in __call__
    raise exc
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\middleware\exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\routing.py", line 443, in handle
    await self.app(scope, receive, send)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\fastapi\applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\middleware\errors.py", line 184, in __call__
    raise exc
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\middleware\errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\middleware\exceptions.py", line 79, in __call__
    raise exc
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\middleware\exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\starlette\routing.py", line 66, in app
    response = await func(request)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\fastapi\routing.py", line 299, in app
    raise e
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\fastapi\routing.py", line 294, in app
    raw_response = await run_endpoint_function(
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\fastapi\routing.py", line 191, in run_endpoint_function
    return await dependant.call(**values)
  File "C:\Users\lucas\python-envs\dashai-dev\lib\site-packages\dependency_injector\wiring.py", line 994, in _patched
    return await _async_inject(
  File "src/dependency_injector/_cwiring.pyx", line 66, in _async_inject
  File "C:\Users\lucas\OneDrive\Desktop\DashAI\DashAI\back\api\api_v1\endpoints\plugins.py", line 256, in update_plugin
    install_plugin_from_pypi(plugin_name)
  File "C:\Users\lucas\OneDrive\Desktop\DashAI\DashAI\back\plugins\utils.py", line 158, in install_plugin_from_pypi
    raise RuntimeError(error_string)
RuntimeError: ERROR: Could not find a version that satisfies the requirement paquete_no_existente (from versions: none)
ERROR: No matching distribution found for paquete_no_existente

Reply:
From what I see, Sqlalchemy is handling the exception. Let's leave it at that for the moment.

Reply:
Ok, so this is ready